model/config.go

```diff
@@ -2501,7 +2525,7 @@ func (ts *TeamSettings) isValid() *AppError {
 }
 
 func (ss *SqlSettings) isValid() *AppError {
-	if len(*ss.AtRestEncryptKey) < 32 {
+	if ss.AtRestEncryptKey != nil && *ss.AtRestEncryptKey != "" && len(*ss.AtRestEncryptKey) < 32 {
```

Comment:
Hmm, this was unexpected, but I think I understand why you added it. The `default.json` should have an empty `AtRestEncryptKey` and `PublicLinkSalt` because, of course, we don't want to pre-ship these keys: they should be generated on server start.

However, the first time a server starts (using the generated `default.json`) and calls `SetDefaults` on /that/ config, it won't generate these values as per the current implementation. I don't think updating `isValid` here is the right path, because the config really /is/ invalid at this point.

What we need to distinguish is the "generate default config.json for package" from "apply defaults on server startup". Fortunately, I think we can do this easily given the use of pointers, e.g. above in `func (s *SqlSettings) SetDefaults`:

```go
    if isUpdate {
        // When updating an existing configuration, ensure an encryption key has been specified.
        if s.AtRestEncryptKey == nil || len(*s.AtRestEncryptKey) == 0 {
 		s.AtRestEncryptKey = NewString(NewRandomString(32))
        }
    } else {
        // When generating a blank configuration, leave this key empty to be generated on server start.
        s.AtRestEncryptKey = NewString("")
    }
```

We'll need to do something similar for `PublicLinkSalt` in the `FileSettings`.

Reply:
@reflog, did we still need to relax the validation here, or can we restore the original checks now?